### ISSUE TITLE
Applying polynomial model using VeLa model creator plugin gives incorrect result

### ISSUE THREAD

**Issue by dbenn** [AAVSO/VStar#227]
Try creating any polynomial fit then applying it via the VeLa model creator plugin.

**Comment by mpyat2** [comment]
1) Load KP Lyn light curve (observer = PMAK) from AAVSO db
2) Polynomial fit degree= 15 for Johnson V
![image](https://user-images.githubusercontent.com/44919945/147731534-881b62c7-4b6d-49ec-bda3-5bec7381d456.png)
3) VeLa model function
zeroPoint is 2459332.35709

f(t:real) : real {
    3.513493389760E19*(t-zeroPoint)^15 +
    1.486885038629E18*(t-zeroPoint)^14 +
    -6.994358939253E17*(t-zeroPoint)^13 +
    -2.778248197535E16*(t-zeroPoint)^12 +
    5.705533522828E15*(t-zeroPoint)^11 +
    2.089863042692E14*(t-zeroPoint)^10 +
    -2.460853674112E13*(t-zeroPoint)^9 +
    -8.103385927143E11*(t-zeroPoint)^8 +
    6.037656131770E10*(t-zeroPoint)^7 +
    1.722865043295E09*(t-zeroPoint)^6 +
    -8.456524341291E07*(t-zeroPoint)^5 +
    -1.993607493521E06*(t-zeroPoint)^4 +
    6.522979829427E04*(t-zeroPoint)^3 +
    1.170687252721E03*(t-zeroPoint)^2 +
    -2.269285161529E01*(t-zeroPoint)^1 +
    1.033057728586E01
}

3) Reload KP Lyn observer PMAK
4) VeLa model for Johnson V
![image](https://user-images.githubusercontent.com/44919945/147731718-597065a2-70f1-4a70-988e-0e6af3619811.png)
4) Model curve
![image](https://user-images.githubusercontent.com/44919945/147731786-0a0f0a25-2280-464b-b15f-59c951946eda.png)

The curves look the same, values (model pane) differ in the 5th digit after the decimal point (it's ok)

I used fresh master brunch

**Comment by dbenn** [comment]
Okay, thanks Max. I'll check that I can reproduce this and if so, see if I can still reproduce the original case I saw this in.

**Comment by dbenn** [comment]
Very strange. I checked out master and then did a git pull. Here's what I see:

![KPLynLCModel](https://user-images.githubusercontent.com/4115628/147846532-3b65c568-177c-49d1-87d8-72366027b9db.png)

I have not done a clean clone. Will do.

**Comment by dbenn** [comment]
I see the same result with a fresh clone on master:
~~~
meeus:vstar david$ git branch -v
* master 157e9a7 Update ReadMe.md
* ~~~

**Comment by mpyat2** [comment]
I pulled code from master right now on the other machine (not my VStar development one), built VStar (and also plig-ins) with ant, and did the test. For me, all works normally.
Only if I select a locale with comma as decimal separator, I could get something similar (yet not the same curve!) because of truncating decimals in this mode (VeLa model used point so the interpreter ignores decimal part if the comma is used).
Also, I tried different time precisions from Preferences and could not reproduce your result too.
It's weird.
VStar 2.22.0dev (git revision 157e9a7f)

**Comment by dbenn** [comment]
Oh, that is very weird. I too wondered whether it might be locale related. Checked that I'm using my default locale (en_AU).

I also tried reversing the latest change to the plugin:

https://github.com/AAVSO/VStar/commit/cb8464b3b5bd4bd5d7b9153a17a709119c1fe66e#diff-d59c1693a01e3bf477d3ee48b0a5c4386b064e874eeb21b3dbe9300a8c07313b

No change. Weird!

I was just looking a `VeLaModelCreator` thinking I should create a plugin test. Might do that and see if it sheds any light on the problem.

**Comment by mpyat2** [comment]
BTW, does the Fourier model (see below) work for you? (I still do not know how it would help, yet just in case...)
I cannot understand why the interpreter gives different results...

---------------
#Fourier model, P= 0.075934, 8 harmonics

zeroPoint is 2459332.357

f(t:real) : real {
10.514335
-0.132323 * cos(2*PI*13.169331*(t-zeroPoint))-0.153611 * sin(2*PI*13.169331*(t-zeroPoint))
-0.045559 * cos(2*PI*26.338663*(t-zeroPoint))-0.052784 * sin(2*PI*26.338663*(t-zeroPoint))
-0.018952 * cos(2*PI*39.507994*(t-zeroPoint))-0.018578 * sin(2*PI*39.507994*(t-zeroPoint))
-0.010631 * cos(2*PI*52.677325*(t-zeroPoint))-0.007972 * sin(2*PI*52.677325*(t-zeroPoint))
-0.006017 * cos(2*PI*65.846656*(t-zeroPoint))-0.005089 * sin(2*PI*65.846656*(t-zeroPoint))
-0.002838 * cos(2*PI*79.015988*(t-zeroPoint))-0.002862 * sin(2*PI*79.015988*(t-zeroPoint))
-0.001407 * cos(2*PI*92.185319*(t-zeroPoint))-0.001839 * sin(2*PI*92.185319*(t-zeroPoint))
-0.000507 * cos(2*PI*105.35465*(t-zeroPoint))-0.000432 * sin(2*PI*105.35465*(t-zeroPoint))
}
------------

**Comment by mpyat2** [comment]
Probably, to isolate the problem, we could start from the simple test: in Tools->VeLa...


zeroPoint is 2459332.35709
f(t:real) : real {
3.513493389760E19*(t-zeroPoint)^15 +
1.486885038629E18*(t-zeroPoint)^14 +
-6.994358939253E17*(t-zeroPoint)^13 +
-2.778248197535E16*(t-zeroPoint)^12 +
5.705533522828E15*(t-zeroPoint)^11 +
2.089863042692E14*(t-zeroPoint)^10 +
-2.460853674112E13*(t-zeroPoint)^9 +
-8.103385927143E11*(t-zeroPoint)^8 +
6.037656131770E10*(t-zeroPoint)^7 +
1.722865043295E09*(t-zeroPoint)^6 +
-8.456524341291E07*(t-zeroPoint)^5 +
-1.993607493521E06*(t-zeroPoint)^4 +
6.522979829427E04*(t-zeroPoint)^3 +
1.170687252721E03*(t-zeroPoint)^2 +
-2.269285161529E01*(t-zeroPoint)^1 +
1.033057728586E01
}

f(2459332.28594)

![image](https://user-images.githubusercontent.com/44919945/147849169-82b57c3b-6c29-463b-b8b0-f01aa72b8a76.png)

**Comment by dbenn** [comment]
> BTW, does the Fourier model (see below) work for you? (I still do not know how it would help, yet just in case...) I cannot understand why the interpreter gives different results...
> 
> #Fourier model, P= 0.075934, 8 harmonics
> 
> zeroPoint is 2459332.357
> 
> f(t:real) : real { 10.514335 -0.132323 * cos(2_PI_13.169331*(t-zeroPoint))-0.153611 * sin(2_PI_13.169331*(t-zeroPoint)) -0.045559 * cos(2_PI_26.338663*(t-zeroPoint))-0.052784 * sin(2_PI_26.338663*(t-zeroPoint)) -0.018952 * cos(2_PI_39.507994*(t-zeroPoint))-0.018578 * sin(2_PI_39.507994*(t-zeroPoint)) -0.010631 * cos(2_PI_52.677325*(t-zeroPoint))-0.007972 * sin(2_PI_52.677325*(t-zeroPoint)) -0.006017 * cos(2_PI_65.846656*(t-zeroPoint))-0.005089 * sin(2_PI_65.846656*(t-zeroPoint)) -0.002838 * cos(2_PI_79.015988*(t-zeroPoint))-0.002862 * sin(2_PI_79.015988*(t-zeroPoint)) -0.001407 * cos(2_PI_92.185319*(t-zeroPoint))-0.001839 * sin(2_PI_92.185319*(t-zeroPoint)) -0.000507 * cos(2_PI_105.35465*(t-zeroPoint))-0.000432 * sin(2_PI_105.35465*(t-zeroPoint)) }

Yep, Fourier is fine.

**Comment by dbenn** [comment]
Indeed, good idea Max re: use of VeLa dialog. The VeLa Model creator plugin test I have in mind will be close to that level too, so good from the viewpoint of regression.

Well, it turns out that I _can_ reproduce your result if I'm running VStar from Eclipse. If I'm building and running via the command line, I get the incorrect result I reported in this issue.

Rather than at first source level debugging `VeLaInterpreter`, I decided to dump the abstract syntax tree (as Lisp-style S-Expressions) generated by executing the example model function code you give. They're different.

In Eclipse:
~~~
(SEQUENCE (IS ZEROPOINT 2459332.35709) (FUNDEF F (PAIR T REAL) REAL (SEQUENCE (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (* 35134933897600000000 (^ (- T ZEROPOINT) 15)) (* 1486885038628999940 (^ (- T ZEROPOINT) 14))) (* -699435893925299970 (^ (- T ZEROPOINT) 13))) (* -27782481975350000 (^ (- T ZEROPOINT) 12))) (* 5705533522828000 (^ (- T ZEROPOINT) 11))) (* 208986304269200 (^ (- T ZEROPOINT) 10))) (* -24608536741120 (^ (- T ZEROPOINT) 9))) (* -810338592714.3 (^ (- T ZEROPOINT) 8))) (* 60376561317.7 (^ (- T ZEROPOINT) 7))) (* 1722865043.295 (^ (- T ZEROPOINT) 6))) (* -84565243.41291 (^ (- T ZEROPOINT) 5))) (* -1993607.493521 (^ (- T ZEROPOINT) 4))) (* 65229.79829427 (^ (- T ZEROPOINT) 3))) (* 1170.687252721 (^ (- T ZEROPOINT) 2))) (* -22.692851615 (^ (- T ZEROPOINT) 1))) 10.330577286))) (F 2459332.28594))
~~~
On the command-line:
~~~
(SEQUENCE (IS ZEROPOINT 2459332.35709) (FUNDEF F (PAIR T REAL) REAL (SEQUENCE (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (+ (* 3.51349339 (^ (- T ZEROPOINT) 15)) (* 1.486885039 (^ (- T ZEROPOINT) 14))) (* -6.994358939 (^ (- T ZEROPOINT) 13))) (* -2.778248198 (^ (- T ZEROPOINT) 12))) (* 5.705533523 (^ (- T ZEROPOINT) 11))) (* 2.089863043 (^ (- T ZEROPOINT) 10))) (* -2.460853674 (^ (- T ZEROPOINT) 9))) (* -8.103385927 (^ (- T ZEROPOINT) 8))) (* 6.037656132 (^ (- T ZEROPOINT) 7))) (* 1.722865043 (^ (- T ZEROPOINT) 6))) (* -8.456524341 (^ (- T ZEROPOINT) 5))) (* -1.993607494 (^ (- T ZEROPOINT) 4))) (* 6.522979829 (^ (- T ZEROPOINT) 3))) (* 1.170687253 (^ (- T ZEROPOINT) 2))) (* -2.269285162 (^ (- T ZEROPOINT) 1))) 1.033057729))) (F 2459332.28594))
~~~
Note for example that we have `3.51349339` in the AST from the command-line above. It's ignoring the exponent part of all real literals, e.g. E09 that's present in the original VeLa function. Whereas, in the Eclipse scenario, we're getting `35134933897600000000` in the AST. That gives the correct result. Source-level debug shows that `35134933897600000000` is being correctly held as a double internally; it's just rendering in the S-Expression AST that way.

So, it will be interesting to know whether you see the same Eclipse vs command-line Ant result. I'm doing this on the master branch.

Also, I just checked: this bug exists in 2.21.3, at least when I launch it on my Mac. Do you see this as well?

Need to understand the difference between those two environments that leads to this discrepancy. Must be something to do with number parsing. I just checked that ANTLR4 via Eclipse and command-line Ant is generating the same Java code. It is.

I also idly wonder whether it relates in any way to #138.

That's where I got to so far.

**Comment by mpyat2** [comment]
Hi @dbenn ,

I've tested the current master on two physical machines (Win7) and virtual Ubuntu installation. All results were successful, VeLa worked properly.
I built the VStar.jar twice using ANT on Machine 1 ( jdk1.8.0_251) and on Machine 3 (jdk-16). To run on Machine 3 I used three different Java versions from command-line (Java 8, 11, 16) and also on Machine 3 I tested VStar from Eclipse that uses JDK 14.

The configurations:

from master: VStar 2.22.0dev (git revision 157e9a7f)

Machine 1 (Windows 7)
fresh ANT build: jdk1.8.0_251

run under
java version "1.8.0_311"
Java(TM) SE Runtime Environment (build 1.8.0_311-b11)
Java HotSpot(TM) 64-Bit Server VM (build 25.311-b11, mixed mode)

Machine 2 (Ubuntu 20.04.3 LTS 64bit in VirtualBox)
the same build as for Machine 1

run under
openjdk version "11.0.13" 2021-10-19
OpenJDK Runtime Environment (build 11.0.13+8-Ubuntu-0ubuntu1.20.04)
OpenJDK 64-Bit Server VM (build 11.0.13+8-Ubuntu-0ubuntu1.20.04, mixed mode, sharing)

Machine 3 (Windows 7: VStar-development-machine)
fresh ANT build: jdk-16 (java version "16" 2021-03-16)

run under three different Java versions from command-line:
 java version "1.8.0_271"
Java(TM) SE Runtime Environment (build 1.8.0_271-b09)
Java HotSpot(TM) 64-Bit Server VM (build 25.271-b09, mixed mode)

java version "11.0.10" 2021-01-19 LTS
Java(TM) SE Runtime Environment 18.9 (build 11.0.10+8-LTS-162)
Java HotSpot(TM) 64-Bit Server VM 18.9 (build 11.0.10+8-LTS-162, mixed mode)

java version "16" 2021-03-16
Java(TM) SE Runtime Environment (build 16+36-2231)
Java HotSpot(TM) 64-Bit Server VM (build 16+36-2231, mixed mode, sharing)

also, from Eclipse on this machine
Eclipse: Java 14, Compiler compliance level 14 (ant uses level 8, probably I should set the level 8 in the Eclipse for the future for compatibility)

**Comment by mpyat2** [comment]
VeLa in VStar 2.21.3 also works normally for me under Windows 7 64bit
java version "1.8.0_311"
Java(TM) SE Runtime Environment (build 1.8.0_311-b11)
Java HotSpot(TM) 64-Bit Server VM (build 25.311-b11, mixed mode)

VStar locale=en

![image](https://user-images.githubusercontent.com/44919945/147854459-8f65271a-49c5-49b0-845b-8f9c81182051.png)

**Comment by mpyat2** [comment]
@dbenn , I found a weird thing.

Look what I did:

Variant 1
a) Set VStar locale = en 
![image](https://user-images.githubusercontent.com/44919945/147856174-d43ded28-7d6d-45b2-bcfa-c146525b752c.png)
b) run under java version "1.8.0_271".
c) VeLa: Correct
![image](https://user-images.githubusercontent.com/44919945/147856249-a58f7280-20a0-41d0-a8c9-5cff4fba4216.png)


Variant 2
a) Set VStar locale = en_AU
b) run under java version "1.8.0_271".
c) VeLa: the same result as in Variant 1: Correct

Variant 3
a) Set VStar locale = en_AU
b) run under java version "16" 2021-03-16 
c)  VeLa: **INCORRECT!!!**
![image](https://user-images.githubusercontent.com/44919945/147856350-82e1b16c-62c8-4f27-a1b4-50ab4a202282.png)

Variant 4
a) Set VStar locale = en (or en_US)
b) run under java version "16" 2021-03-16  again
c) VeLa: CORRECT AGAIN!!!
![image](https://user-images.githubusercontent.com/44919945/147856385-2162c867-1e2d-4d52-a87b-a320714265ff.png)

So, under Java 16, locale en_AU gives incorrect results, while en_US -- correct!!!

Well, I also tried different European locales with ',' as a decimal separator.

Under Java 16
VeLa 
a is 3,141592E5
a

evaluated as 3,141592 if locale = uk_UA (Ukraine): incorrect, however it is evaluated correctly if locale = de_DE (Germany)

However, under Java 8 uk_UA works correctly (the result = 314159,2)

New Zealand locale en_NZ works correctly for both Java 8 and Java 16.

en_AU locale also gives incorrect results for Java 11 (the result = 3.141592).

Well, I have no idea why Australian (and Ukrainian, as I see) locales are so strange for Java ver >8.

It would be interesting to know what Java version are you using in the command line and what in Eclipse.

**Comment by mpyat2** [comment]
Ubuntu, openjdk version "11.0.13" 2021-10-19
locale=en_AU: exponent is ignored
locale=en_US: correct

**Comment by mpyat2** [comment]
I also made a little java program, that takes a string from the console and parses it using Number format:
<...>
NumberFormat numberFormat = NumberFormat.getNumberInstance(Locale.getDefault());
Number number = numberFormat.parse(s);
double d = number.doubleValue();
System.out.println(d);
<...>

Running it using different locales (with java 16!) (for example "java -Duser.country=AU -Duser.language=en Test") I found that numberFormat.parse correctly parses low-case exponent for en_AU (for example, 3.14159e5) yet cannot parse 3.14159E5.

This does not work for VeLa, though.

java 1.8 works differently,

**Comment by dbenn** [comment]
Thanks for the exhaustive locale checking @mpyat2!

Okay, well my command-line Java is currently:
~~~
meeus:vstar david$ java -version
openjdk version "11.0.10" 2021-01-19 LTS
OpenJDK Runtime Environment Corretto-11.0.10.9.1 (build 11.0.10+9-LTS)
OpenJDK 64-Bit Server VM Corretto-11.0.10.9.1 (build 11.0.10+9-LTS, mixed mode)
meeus:vstar david$ ant
Buildfile: /Users/david/Projects/working_copy/VStar/build.xml
~~~
So, OpenJDK 11, similar to your example Ubuntu version above.

I just changed the locale to `en_US` and it works perfectly for the simple `a is 3.141592E5` and the KP Lyn model.

I can also confirm that `uk_UA` and `en_AU` _both fail to yield correct results_, consistent with your findings.

My current Eclipse Java version is 1.8.0_201 (compliance level set to 1.8) Both `uk_UA` and `en_AU` work correctly with this version.

So, there we have it. Sometimes Java really isn't "write once, run anywhere" eh?

It's worse than that: it seems to be targeting you and I, or at least our countries! ;)

Approaches to solving this?
* Detect `en_AU` and `uk_UA` at run time and set to `en_US` or something else?
* Run a little VeLa test of the sort you used above (`a is 3.141592E5`) at startup time and if it fails, set a "safe" locale such as `en_US`? This should be fairly low cost.
* Document what has been found here re: Java versions and point people to this when necessary?

**Comment by dbenn** [comment]
As per last commit, I've added some unit tests that test the code we've been experimenting with here for all available locales with whatever Java version we happen to be running.

Currently, our GitHub CI actions system used Java 1.8 (JAVA_HOME: /opt/hostedtoolcache/jdk/8.0.312/x64) and the tests pass there but we could add more Java version runners.

**Comment by mpyat2** [comment]
Hi David,

The approach with a little VeLa test at startup would be a good solution ("safe" locale can be "en" without country code).

However, after some thought and a little experimenting, I see another possible solution.

We could change the code of ExpressionVisitor.parseDouble().

Instead of 
```
NumberFormat FORMAT = NumberFormat.getNumberInstance(Locale
				.getDefault());
```

use something like this:

```
DecimalFormatSymbols dfs = new DecimalFormatSymbols();
dfs.setExponentSeparator("E");
DecimalFormat FORMAT = new DecimalFormat("0.000000000000000E0", dfs);
```

![image](https://user-images.githubusercontent.com/44919945/147887262-3d0bc7eb-b99d-4f7c-965e-24606cdeae45.png)

It seems this solves the problem with our locales (are we the only exceptions?). Well, additional tests are needed, probably the DecimalFormat pattern is not optimal (the documentation should be checked) yet it apparently works (at least for me).
Maybe DecimalFormatSymbols(Locale.getDefault()) would be better (or at least clearer, I don't know)

New Java libraries (at least in Java 11, 14, 16) use low-case 'e' as an Exponent Separator (who knows why?) for en_AU instead of capital E.
For the Ukrainian locale, uk_UA, the Exponent Separator is a Cyrillic 'E' (looks exactly like the Latin equivalent, the pronunciation like the first sound in the word 'element'), unicode = 0415 (at least in Java 11, 14, 16). Honestly saying, it's a surprise for me, I have never seen any software product that uses Cyrillic 'E' instead of Latin 'E' or 'e' as an exponent separator. Moreover, Java 8 uses the Latin E for Cyrillic locales!

Such implementation of locales looks quite strange (it's a polite word, I don't want to write what I really think about this).

**Comment by dbenn** [comment]
Hi @mpyat2 Again, thanks. I like your solution better than mine. Please go ahead, if you are happy to!

The unit tests I added and committed yesterday will help pick up any problems with this approach. I will also add to the GitHub CI YAML to test with more versions of Java as well.

**Comment by mpyat2** [comment]
See Issue #229

**Comment by dbenn** [comment]
The `VeLaTest` unit tests pass with your proposed code @mpyat2 in case you hadn't tried it. As you say, the exact format may need some refinement, but definitely getting there.

**Comment by mpyat2** [comment]
@dbenn , thank you very much for the tests! I made additional changes and opened a pool request.

**Comment by dbenn** [comment]
@mpyat2 Now that you've fixed the issue in #230 and that branch is merged, I'll create a pull request for you to review the unit tests, merge this branch and then add multiple Java version handling to GitHub CI via #120, which I'll then ask you to review. Does that all seem OK?

**Comment by dbenn** [comment]
@mpyat2 The combination of #230 and #231 (expression visitor fix and numeric locale UTs) merged to master is good. VeLaTest.java passes.

Given the separation of VStar core UTs and plugin UTs on #120, I'll add multiple Java versions to the CI x 2 YAML files on that issue's current branch. I should get that finished in the next day or so.
